x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/threat_marker/scan.ts

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import type { Client } from '@elastic/elasticsearch';
+import async from 'async';
+
+export interface Threat {
+  'threat.indicator.type': string;
+  'threat.indicator.url.full': string;
+  'threat.indicator.file.hash.md5': string;
+  'threat.indicator.file.hash.sha1': string;
+}
+
+import type {
+  OpenPointInTimeResponse,
+  SortResults,
+  QueryDslQueryContainer,
+  SearchHit,
+} from '@elastic/elasticsearch/lib/api/types';
+
+/**
+ * Returns filter clause with 'match' conditions for relevant fields only
+ * @param threat
+ * @returns
+ */
+export const shouldClauseForThreat = (threat: Threat) => {
+  switch (threat['threat.indicator.type']) {
+    case 'url': {
+      return [
+        {
+          match: {
+            'url.full': threat['threat.indicator.url.full'],
+          },
+        },
+      ];
+    }
+
+    case 'file': {
+      return [
+        {
+          match: {
+            'file.hash.md5': threat['threat.indicator.file.hash.md5'],
+            'file.hash.sha1': threat['threat.indicator.file.hash.sha1'],
+          },
+        },
+      ];
+    }
+  }
+};
+
+export const THREAT_DETECTION_INDICATOR_FIELD = 'threat.detection.indicator' as const;
+
+export const THREAT_DETECTION_TIMESTAMP_FIELD = 'threat.detection.timestamp' as const;
+
+const updateMapping = async (client: Client, eventsIndex: string[]) => {
+  await client.indices.putMapping({
+    index: eventsIndex,
+    properties: {
+      [THREAT_DETECTION_INDICATOR_FIELD]: {
+        type: 'keyword',
+      },
+      [THREAT_DETECTION_TIMESTAMP_FIELD]: {
+        type: 'date',
+      },
+    },
+  });
+};
+
+export const mark = async (
+  es: Client,
+  index: string[],
+  query: QueryDslQueryContainer,
+  indicator: string,
+  timestamp: number
+) =>
+  es.updateByQuery({
```

Comment:
am I correct that the main difference with the Indicator Match Rule is in this logic? That this rule doesn't create its own artifacts and therefore doesn't write to Elasticsearch, only updates.

Reply:
Yes, but It also happens in scope of a single ES request and its up to the search engine to carry on the query and potential upates if matches are found.